lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -167,18 +191,51 @@ bool ArrayAllocation::isInitializationWithKnownElements() {
 ///
 /// We have to prove that the array value is not changed in between the
 /// creation and the method call to get_element.
-bool ArrayAllocation::findValueReplacements() {
+bool ArrayAllocation::findReplacements() {
   if (!isInitializationWithKnownElements())
     return false;
 
   // The array value was stored or has escaped.
   if (!analyseArrayValueUses())
     return false;
 
-  // No count users.
-  if (GetElementCalls.empty())
+  if (GetElementCalls.empty() && AppendContentsOfCalls.empty())
     return false;
 
+  if (!AppendContentsOfCalls.empty()) {
+    // Figure out the formal type of the array's generic parameter
+    auto SwiftTy = dyn_cast<TupleType>(Alloc->getType().getSwiftType());
+    auto ElTy = dyn_cast<BoundGenericStructType>(SwiftTy.getElementType(0));
+    auto GenericArgs = ElTy.getGenericArgs();
+    assert(GenericArgs.size() == 1 && "There should be a generic parameter");
+    auto FormalTy = GenericArgs[0];
+
+    // Create a vector from the map
+    bool ReplacementsAreValid = true;
+    auto ElementCount = ElementValueMap.size();
+    llvm::SmallVector<SILValue, 4> ElementValueVector;
+
+    if (ElementCount > APPEND_CONTENTSOF_REPLACEMENT_VALUES_MAX) {
+      ReplacementsAreValid = false;
+    }
+    else {
+      for (size_t i = 0; i < ElementCount; ++i) {
+        if (!ElementValueMap.count(i)) {
+          // This means that elements aren't contiguous
+          ReplacementsAreValid = false;
+          break;
+        }
+
+        ElementValueVector.push_back(ElementValueMap[i]);
+      }
+    }
+
+    if (ReplacementsAreValid)
+      for (auto *Call : AppendContentsOfCalls)
```

Comment:
If you do what I suggest above, you can reduce indentation here as well. So then you get:

if (!ReplacementsAreValid) {
    return collectForwardableValues();
}

for (auto *Call : ...) {

}